File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> st = new Stack<>();
        st.push(s.charAt(0));
        for(int i=1; i<s.length(); i++) {
            char c = s.charAt(i);
            if(c == ')' || c == '}' || c == ']' ) {
                if(st.isEmpty()) return false;
                if(st.peek() == '(' && c == ')' || st.peek() == '{' && c == '}' || st.peek() == '[' && c == ']') {
                    st.pop();
                }
                else return false;
            }
            else {
                st.push(c);
            }

        }
        return st.isEmpty();
    }
}

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        st = []
        st.append(s[0])

        for i in range(1, len(s)):
            c = s[i]
            if c in [')', '}', ']']:
                if not st: return False
                if(st[-1] == '(' and c == ')' or st[-1] == '{' and c == '}' or st[-1] == '[' and c == ']'):
                    st.pop()
                else:
                    return False
            else:
                st.append(c)

        return len(st)==0